src/data/data.cc

```diff
@@ -39,15 +122,42 @@ void MetaInfo::Clear() {
   base_margin_.HostVector().clear();
 }
 
+/*
+ * Used format:
+ *
+ * | name        | type     | rows    | columns | value           |
+ * |-------------+----------+---------+---------+-----------------|
+ * | num_row     | kUInt64  | NA      |      NA | ${num_row_}     |
+ * | num_col     | kUInt64  | NA      |      NA | ${num_col_}     |
+ * | num_nonzero | kUInt64  | NA      |      NA | ${num_nonzero_} |
+ * | labels      | kFloat32 | ${size} |       1 | ${labels_}      |
+ * | group_ptr   | kUInt32  | ${size} |       1 | ${group_ptr_}   |
+ * | weights     | kFloat32 | ${size} |       1 | ${weights_}     |
+ * | base_margin | kFloat32 | ${size} |       1 | ${abse_margin}  |
+ *
+ * Notice the difference between saved name and the name used in `SetInfo': the former
+ * uses plural.
+ */
+
 void MetaInfo::SaveBinary(dmlc::Stream *fo) const {
   Version::Save(fo);
-  fo->Write(&num_row_, sizeof(num_row_));
-  fo->Write(&num_col_, sizeof(num_col_));
-  fo->Write(&num_nonzero_, sizeof(num_nonzero_));
-  fo->Write(labels_.HostVector());
-  fo->Write(group_ptr_);
-  fo->Write(weights_.HostVector());
-  fo->Write(base_margin_.HostVector());
+  fo->Write(kNumField);
+  int field_cnt = 0;  // make sure we are actually writing kNumField fields
+
+  SaveScaleField(fo, u8"num_row", DataType::kUInt64, num_row_); ++field_cnt;
```

Comment:
I think we should include a flag to indicate whether a field is scalar or vector, so that the binary format is self-describing.

Reply:
Yup.  That sounds great.